src/flashcart/ed64/ed64_internal.c

```diff
@@ -0,0 +1,690 @@
+
+    volatile uint32_t piLatReg = IO_READ(PI_BSD_DOM2_LAT_REG);
+    volatile uint32_t piPwdReg = IO_READ(PI_BSD_DOM2_PWD_REG);
+    volatile uint32_t piPgsReg = IO_READ(PI_BSD_DOM2_PGS_REG);
+    volatile uint32_t piRlsReg = IO_READ(PI_BSD_DOM2_RLS_REG);
+
+    IO_WRITE(PI_BSD_DOM2_PGS_REG, 0x0D);
+    IO_WRITE(PI_BSD_DOM2_RLS_REG, 0x02);
+    IO_WRITE(PI_BSD_DOM2_LAT_REG, 0x05);
+    IO_WRITE(PI_BSD_DOM2_PWD_REG, 0x0C);
+
+    ed64_bios_dma_read(ram, SRAM_ADDRESS + address, length);
+
+    IO_WRITE(PI_BSD_DOM2_LAT_REG, piLatReg);
+    IO_WRITE(PI_BSD_DOM2_PWD_REG, piPwdReg);
+    IO_WRITE(PI_BSD_DOM2_PGS_REG, piPgsReg);
+    IO_WRITE(PI_BSD_DOM2_RLS_REG, piRlsReg);
+}
+
+/* Write to SRAM over DMA */
+void ed64_bios_dma_write_sram(void *ram, uint32_t address, uint32_t length) {
+
+    volatile uint32_t piLatReg = IO_READ(PI_BSD_DOM2_LAT_REG);
+    volatile uint32_t piPwdReg = IO_READ(PI_BSD_DOM2_PWD_REG);
+    volatile uint32_t piPgsReg = IO_READ(PI_BSD_DOM2_PGS_REG);
+    volatile uint32_t piRlsReg = IO_READ(PI_BSD_DOM2_RLS_REG);
+
+    IO_WRITE(PI_BSD_DOM2_PGS_REG, 0x0D);
+    IO_WRITE(PI_BSD_DOM2_RLS_REG, 0x02);
+    IO_WRITE(PI_BSD_DOM2_LAT_REG, 0x05);
+    IO_WRITE(PI_BSD_DOM2_PWD_REG, 0x0C);
+
+    ed64_bios_dma_write(ram, SRAM_ADDRESS + address, length);
+
+
+    IO_WRITE(PI_BSD_DOM2_LAT_REG, piLatReg);
+    IO_WRITE(PI_BSD_DOM2_PWD_REG, piPwdReg);
+    IO_WRITE(PI_BSD_DOM2_PGS_REG, piPgsReg);
+    IO_WRITE(PI_BSD_DOM2_RLS_REG, piRlsReg);
+
+}
+
+uint16_t ed64_bios_msg_rd() {
+
+    return ed64_bios_reg_read(REG_MSG);
+}
+
+void ed64_bios_msg_wr(uint16_t val) {
+
+    ed64_bios_reg_write(REG_MSG, val);
+}
+
+typedef struct PI_regs_s {
+    /** @brief Uncached address in RAM where data should be found */
+    void * ram_address;
+    /** @brief Address of data on peripheral */
+    unsigned long pi_address;
+    /** @brief How much data to read from RAM into the peripheral */
+    unsigned long read_length;
+    /** @brief How much data to write to RAM from the peripheral */
+    unsigned long write_length;
+    /** @brief Status of the PI, including DMA busy */
+    unsigned long status;
+} _PI_regs_s;
+
+static volatile struct PI_regs_s * const PI_regs = (struct PI_regs_s *) PI_BASE_REG;
+
+void ed64_bios_dma_pi_read(void * ram_address, unsigned long pi_address, unsigned long length) {
+
+    disable_interrupts();
+    dma_wait();
+    IO_WRITE(PI_STATUS_REG, 3);
+    PI_regs->ram_address = ram_address;
+    PI_regs->pi_address = pi_address & 0x1FFFFFFF;
+    PI_regs->write_length = length - 1;
+    dma_wait();
+
+    enable_interrupts();
+}
+
+void ed64_bios_dma_pi_write(void * ram_address, unsigned long pi_address, unsigned long length) {
+
+    disable_interrupts();
+    dma_wait();
+    IO_WRITE(PI_STATUS_REG, 3);
+    PI_regs->ram_address = ram_address;
+    PI_regs->pi_address = pi_address & 0x1FFFFFFF;
+    PI_regs->read_length = length - 1;
+    dma_wait();
+    enable_interrupts();
+}
```

Comment:
Same as with `io_read`/`io_write`, why not use `dma_read_async`/`dma_write_raw_async` instead?